dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/validation/hooks/EnrollmentAttributeValidationHook.java

```diff
@@ -0,0 +1,188 @@
+            }
+
+            attributeValueMap.put( attribute.getAttribute(), attribute.getValue() );
+
+            validateAttrValueType( reporter, attribute, teAttribute );
+
+            //NOTE: this is perf killing
+            validateAttributeUniqueness( reporter,
+                attribute.getValue(),
+                teAttribute,
+                tei,
+                tei.getOrganisationUnit() );
+        }
+
+        validateMandatoryAttributes( reporter, program, tei, attributeValueMap );
+    }
+
+    protected void validateRequiredProperties( ValidationErrorReporter reporter, Attribute attribute )
+    {
+        if ( attribute.getAttribute() == null )
+        {
+            reporter.addError( newReport( TrackerErrorCode.E1075 )
+                .addArg( attribute ) );
+        }
+
+        if ( attribute.getValue() == null )
+        {
+            reporter.addError( newReport( TrackerErrorCode.E1076 )
+                .addArg( attribute ) );
+        }
+
+        if ( attribute.getAttribute() != null )
+        {
+            TrackedEntityAttribute teAttribute = reporter.getValidationContext()
+                .getTrackedEntityAttribute( attribute.getAttribute() );
+            if ( teAttribute == null )
+            {
+                reporter.addError( newReport( TrackerErrorCode.E1017 )
+                    .addArg( attribute.getAttribute() ) );
+            }
+        }
+    }
+
+    private void validateMandatoryAttributes( ValidationErrorReporter reporter,
+        Program program, TrackedEntityInstance trackedEntityInstance, Map<String, String> attributeValueMap )
+    {
+        Objects.requireNonNull( program, TrackerImporterAssertErrors.PROGRAM_CANT_BE_NULL );
+        Objects.requireNonNull( trackedEntityInstance, TRACKED_ENTITY_INSTANCE_CANT_BE_NULL );
+        Objects.requireNonNull( attributeValueMap, ATTRIBUTE_VALUE_MAP_CANT_BE_NULL );
+
+        // 1. Get all tei attributes, map attrValue attr. into set of attr.
+        Set<TrackedEntityAttribute> trackedEntityAttributes = trackedEntityInstance.getTrackedEntityAttributeValues()
+            .stream()
+            .map( TrackedEntityAttributeValue::getAttribute )
+            .collect( Collectors.toSet() );
+
+        // 2. Map all program attr. that match tei attr. into map. of attr:is mandatory
+        Map<TrackedEntityAttribute, Boolean> mandatoryMap = program.getProgramAttributes().stream()
+            .filter( v -> trackedEntityAttributes.contains( v.getAttribute() ) )
+            .collect( Collectors.toMap(
+                ProgramTrackedEntityAttribute::getAttribute,
+                ProgramTrackedEntityAttribute::isMandatory ) );
+
+        for ( Map.Entry<TrackedEntityAttribute, Boolean> entry : mandatoryMap.entrySet() )
+        {
+            TrackedEntityAttribute attribute = entry.getKey();
+            Boolean attributeIsMandatory = entry.getValue();
+
+            // TODO: This is quite ugly and should be considered to be solved differently,
```

Comment:
I agree. leave it as it is for now, then we need to figure out the usecase for this and see if we can solve it differently.